internal/dataplane/sendconfig/inmemory_error_handling.go

```diff
@@ -111,85 +110,106 @@ func parseFlatEntityErrors(body []byte, logger logr.Logger) ([]ResourceError, er
-				if len(message) > 0 {
-					raw.Problems[fmt.Sprintf("%s[%d]", p.Field, i)] = message
-				}
-			}
-		}
+		raw := flattenedErrorIntoRaw(ee, logger)
 		parsed, err := parseRawResourceError(raw)
 		if err != nil {
-			logger.Error(err, "Entity tags missing fields", "name", ee.Name)
+			logger.Error(err, "Entity tags returned in error API error from gateway is missing fields, Kubernetes Event cannot be created without them",
+				"name", ee.Name,
+				"id", ee.ID,
+				"tags", ee.Tags,
+				"type", ee.Type,
+				"raw_error", ee,
+			)
 			continue
 		}
 		resourceErrors = append(resourceErrors, parsed)
 	}
 	return resourceErrors, nil
 }
 
+func flattenedErrorIntoRaw(ee FlatEntityError, logger logr.Logger) rawResourceError {
+	raw := rawResourceError{
+		Name:     ee.Name,
+		ID:       ee.ID,
+		Tags:     ee.Tags,
+		Problems: map[string]string{},
+	}
+	for _, p := range ee.Errors {
+		if len(p.Message) > 0 && len(p.Messages) > 0 {
+			logger.Error(nil, "Entity has both single and array errors for field",
+				"name", ee.Name, "field", p.Field)
+			continue
+		}
+		if len(p.Message) > 0 {
+			switch p.Type {
+			case FlatErrorTypeField:
+				// If the error is associated with a single field, store it in the map under the field name.
+				raw.Problems[p.Field] = p.Message
+			case FlatErrorTypeEntity:
+				// If the error is associated with a whole entity, store it in the map under the entity type and name.
+				raw.Problems[fmt.Sprintf("%s:%s", ee.Type, ee.Name)] = p.Message
+			}
+		}
+		for i, message := range p.Messages {
+			if len(message) > 0 {
+				raw.Problems[fmt.Sprintf("%s[%d]", p.Field, i)] = message
+			}
+		}
+	}
+	return raw
+}
+
 // parseRawResourceError takes a raw resource error and parses its tags into Kubernetes metadata. If critical tags are
 // missing, it returns an error indicating the missing tag.
 func parseRawResourceError(raw rawResourceError) (ResourceError, error) {
-	re := ResourceError{}
-	re.Problems = raw.Problems
+	re := ResourceError{
+		Problems: raw.Problems,
+	}
+
 	var gvk schema.GroupVersionKind
 	for _, tag := range raw.Tags {
 		if strings.HasPrefix(tag, util.K8sNameTagPrefix) {
 			re.Name = strings.TrimPrefix(tag, util.K8sNameTagPrefix)
+			continue
 		}
 		if strings.HasPrefix(tag, util.K8sNamespaceTagPrefix) {
 			re.Namespace = strings.TrimPrefix(tag, util.K8sNamespaceTagPrefix)
+			continue
 		}
 		if strings.HasPrefix(tag, util.K8sKindTagPrefix) {
 			gvk.Kind = strings.TrimPrefix(tag, util.K8sKindTagPrefix)
+			continue
 		}
 		if strings.HasPrefix(tag, util.K8sVersionTagPrefix) {
 			gvk.Version = strings.TrimPrefix(tag, util.K8sVersionTagPrefix)
+			continue
 		}
 		// this will not set anything for core resources
 		if strings.HasPrefix(tag, util.K8sGroupTagPrefix) {
 			gvk.Group = strings.TrimPrefix(tag, util.K8sGroupTagPrefix)
+			continue
 		}
 		if strings.HasPrefix(tag, util.K8sUIDTagPrefix) {
 			re.UID = strings.TrimPrefix(tag, util.K8sUIDTagPrefix)
+			continue
 		}
```

Comment:
Maybe turning this into a switch would make sense, WDYT?

Reply:
👍 PTAL 174a92af8db0724190645739cc50bf6118708579